### DISCUSSION TITLE
Seperate python2 manifests from the python3 ones

### DISCUSSION THREAD

**Discussion by ardenercelik** [microsoft/winget-pkgs#12004] [General]
Currently python2 and python3 packages are located in the same folder. I would like to install the latest python2 release without specifying the version by typing "winget install python2" or "winget install python27"

**Comment by OfficialEsco** [comment]
I've been thinking about this for a few other apps too, technically the `PackageName` for a few Packages are like `Python 2.7.9 (x64)` and `Python 3.9.2 (x64)` therefore if we fix the PackageName for those programs you should theoretically be able to do `winget install "Python 2"` and it will install the latest version of `Python 2.*.*`

However as you can see there is another problem since the `x64` and `x86` version have 2 different `PackageNames`, so we would need 2 different manifest eitherway, BUT ALSO this would bring more issues to the PR submissions where bots don't check for `PackageName` and people JUST update the `InstallerURL`, `Hash` and `Version`...

Yes its possible but the PR submissions would have to be stricter 🙄 

@denelon

**Comment by jedieaston** [comment]
> However as you can see there is another problem since the `x64` and `x86` version have 2 different `PackageNames`

Someone (I can do it unless someone gets to it before me) needs to go through and remove the architectures from the `PackageName`, since I added x86 architectures to all of the Python manifests a while ago and didn't realize that was there. Oops.

But we can move the Python2 ones to Python.Python.2 and the Python3 ones to Python.Python.3 (or keep them in Python.Python, since Python2 is EOL and people need to start their migrations... ;D). That's what I did for OpenJDK, and it works quite well.

Anyone have a preference?

**Comment by denelon** [comment]
There are several opinions about versioning and programming languages. Now that we have solid support for multi-architecture manifests, it is possible to merge packages with the same version into a single logical package. As the Windows Package Manager doesn't support pinning for versions yet, it seems logical to have different packages for different versions. Once we have pinning it would be possible to collapse again into a single package. I'd say it's more of a community preference.

This might be a great area for a discussion.

**Comment by jedieaston** [comment]
> This might be a great area for a discussion.

Yes please.

**Comment by denelon** [comment]
I've converted the Issue to a discussion.

I also pinned this discussion to get more visibility.

https://twitter.com/DenelonMs/status/1389626501742489605

**Comment by jedieaston** [comment]
I've just moved all of the Python 2 manifests to `PackageIdentifier` Python.Python.2 (since there was no further discussion, I figured everyone was fine with it. I can move everything back if someone has a problem with it, but the bot seemed to be okay with deleting the Python.Python manifests before adding the new ones, so....) 

Presuming all of these clear, I'll probably move the Python 3.x manifests to Python.Python.3 and add a `Moniker: python` so they are the "default" ones. (Again, unless someone has a better idea, but this keeps it in line with OpenJDK and probably how .NET Core will be when I or someone else gets around to it).

**Comment by jedieaston** [comment]
Alrighty, looks like this is all done once the PRs clear (the pipeline got stuck on a couple). Within the next hour or two, you should be able to do `winget install Python.Python.2` and `winget install Python.Python.3`! 🎉

**Reply by denelon** [reply]
I noticed the latest version is 3.9.4150.0 and not 3.9.5. Did something strange happen with their releases recently?

**Reply by jedieaston** [reply]
That's a patch to version 3.9.4 (and they replaced the 3.9.4 binary on their server with that one). I don't know why they gave it a weird version number, but 3.9.5 is the latest version.

**Reply by denelon** [reply]
That ends up breaking as the number 4150 is greater in magnitude than 5. It's not [semantic](https://semver.org) 🙁. If I try to install the latest version, it will give me the 4150 version. I wonder if it would be better to call it 3.9.4.4150 or something else.

**Reply by jedieaston** [reply]
It looks like 3.9.5 shows up in the ARP table as 3.9.5150.0 too, I'll make a corrected manifest and open a PR.

![image](https://user-images.githubusercontent.com/21368066/118166726-d5c7d480-b3f3-11eb-9f3a-123e0d9b8336.png)

**Reply by jedieaston** [reply]
It has been done.

```
PS C:\Users\Easton Pillay\projects\winget-pkgs> winget show Python.Python.3
Found Python 3.9.5 [Python.Python.3]
Version: 3.9.5150.0
Publisher: Python
Moniker: python
Description: Python is a programming language that lets you work more quickly and integrate your systems more effectively.
Homepage: https://www.python.org
License: PSF LICENSE AGREEMENT FOR PYTHON
License Url: https://docs.python.org/3/license.html
Installer:
  Type: Exe
  Download Url: https://www.python.org/ftp/python/3.9.5/python-3.9.5-amd64.exe
  SHA256: 84d5243088ba00c11e51905c704dbe041040dfff044f4e1ce5476844ee2e6eac
PS C:\Users\Easton Pillay\projects\winget-pkgs> # Hooray!
```

**Comment by denelon** [comment]
I'd say the Python community owes you one 🥇
